HAFS related moving-nesting, nest coupling, and movable output grid capabilities

BinLiu-NOAA opened the pull request:
## Description

This pull request includes the UFS/HAFS related moving-nesting developments (including storm-following or specified moving nesting for the regional or global configuration, HYCOM ocean coupling for the moving nesting configuration, write grid component support for moving nesting). More detailed description and list of contributors can be found in ufs-community/ufs-weather-model/pull/1104.

### Issue(s) addressed

Link the issues to be closed with this PR, whether in this repository, or in another repository.
(Remember, issues should always be created before starting work on a PR branch!)
- addresses ufs-community/ufs-weather-model/issues/1100
- addresses ufs-community/ufs-weather-model/issues/1053

## Testing

Five HAFS moving nesting related regression tests are added in ufs-weather-model. And the full regression tests will be conducted in ufs-weather-model before the merge.

## Dependencies

- associated with ufs-community/ufs-weather-model/pull/1104
- waiting on NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/179
- depend upon ufs-community/ufs-weather-model/pull/1135
- depend upon ufs-community/ufs-weather-model/pull/1146

climbfuji committed:
Update .gitmodules and submodule pointer for ccpp-physics for code review and testing

climbfuji committed:
Merge branch 'support/HAFS' of https://github.com/hafs-community/fv3atm into update_dtc_hwrf_physics_from_hafs_community_and_fix_FA_table_initialization

climbfuji committed:
Merge branch 'dtc_hwrf_physics' of https://github.com/mzhangw/fv3atm into update_dtc_hwrf_physics_from_hafs_community_and_fix_FA_table_initialization

climbfuji committed:
Update submodule pointer for ccpp/physics

climbfuji committed:
Revert change to .gitmodules and update submodule pointer for ccpp-physics

climbfuji committed:
Merge pull request #50 from climbfuji/update_dtc_hwrf_physics_from_hafs_community_and_fix_FA_table_initialization

dtc/hwrf-physics: update from HAFS community and SDFs for HWRF physics test

grantfirl committed:
Merge branch 'dtc/hwrf-physics' into hwrf_sfc_PR

climbfuji committed:
Update .gitmodules and submodule pointer for code review and testing

grantfirl committed:
Merge pull request #1 from climbfuji/hwrf_sfc_PR_dom_for_grant

Update .gitmodules and submodule pointer for code review and testing

grantfirl committed:
update submodule pointer for ccpp/physics

grantfirl committed:
update submodule pointer for ccpp/physics and revert changes to .gitmodules

grantfirl committed:
Merge pull request #49 from grantfirl/hwrf_sfc_PR

support HWRF Noah LSM and GFDL surface layer

BinLiu-NOAA committed:
Add a new ccpp suite of suite_HAFS_v0_gfdlmp_nocpugwd.xml.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_couplehycom

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

danrosen25 committed:
Merge NUOPC sst import with internal sfcprop%tsfc if cell contains
missing values.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_couplehycom

climbfuji committed:
Merge branch 'support/HAFS' of https://github.com/hafs-community/fv3atm into HEAD

climbfuji committed:
Update .gitmodules and submodule pointers for GFDL_atmos_cubed_sphere, ccpp-framework, ccpp-physics

mzhangw committed:
modify HWRF_nogwdps as HWRF physics test

climbfuji committed:
Merge branch 'nogwdps_sdf' of https://github.com/mzhangw/fv3atm into update_dtc-hwrf-physics_from_hafs_support_20200618

climbfuji committed:
Correct CCPP version number in HWRF SDFs

climbfuji committed:
Revert change to .gitmodules and update submodule pointer for ccpp-physics

climbfuji committed:
Merge pull request #54 from climbfuji/update_dtc-hwrf-physics_from_hafs_support_20200618

dtc/hwrf-physics: update from support/hafs 2020/06/18

These code changes have been approved previously when merged into the authoritative branches (which were then brought to support/HAFS). Merging without approval.

climbfuji committed:
Update .gitmodules and submodule pointer for ccpp-physics for code review and testing

BinLiu-NOAA committed:
Update the ccpp/physics sumodule to point to the feature/hwrf_physics branch in the hafs-community fork.

BinLiu-NOAA committed:
Add a new SDF suite_HAFS_v0_gfdlmp_nocpnsst.xml by turning off nsst.

BinLiu-NOAA committed:
Add the sfc_ocean item in suite_HAFS_v0_gfdlmp_nocpnsst.xml.

BinLiu-NOAA committed:
Rename or clean up some HAFS related ccpp suite definition files.

climbfuji committed:
Merge pull request #3 from hafs-community/feature/merge_hwrf_physics

Rename or clean up some HAFS hwrf physics related SDFs

climbfuji committed:
Update .gitmodules and submodule pointer for ccpp-physics for code review and testing

climbfuji committed:
Remove suite_FV3_HAFS_FA_HWRF_RRTMG.xml

climbfuji committed:
Revert change to .gitmodules and update submodule pointer for ccpp-physics

climbfuji committed:
Merge pull request #5 from climbfuji/update_hafs_from_dtc_hwrf_physics_20200619

Update support/HAFS from NCAR dtc/hwrf-physics 2020/06/19

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Following the changes in other SDFs, move GFS_DCNV_generic_pre to the proper
location for HAFS related SDFs.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_couplehycom
And add a new SDF of suite_HAFS_v0_gfdlmp_nocpnsstugwd.xml

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/sync_develop_20200703

BinLiu-NOAA committed:
Update submodule ccpp/physics

BinLiu-NOAA committed:
Update submodule ccpp/physics

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/sync_develop_20200703' into feature/hafs_couplehycom

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Udpate the atmos_cubed_sphere submodule to point to the support/HAFS branch in
the hafs-community fork, which has a bug fix to enable running
ufs-weather-model with other vertical level configurations (in addition to
L64), when using the external initial conditions generated by UFS_UTILS from
GFS input files.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/sync_develop_20200703

BinLiu-NOAA committed:
Update ccpp/physics.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/sync_develop_20200703' into feature/hafs_couplehycom

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge branch 'support/HAFS' of https://github.com/hafs-community/fv3atm into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Add a new HAFS suite: suite_HAFS_v0_gfdlmp_nonsstugwd.xml, turning on the convection.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

danrosen25 committed:
Add corners to regional grid with gridAlign -1,-1. Add write_grid_netcdf
for diagnostics.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

wramstrom committed:
Initial Moving Nest Functionality

BinLiu-NOAA committed:
Merge branch 'support/HAFS' into feature/hafs_moving_nest

BinLiu-NOAA committed:
Add a new ccpp SDF: suite_HAFS_v0_gfdlmp_nonsst.xml

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

Conflicts:
	ccpp/physics

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere

BinLiu-NOAA committed:
Merge branch 'support/HAFS' into feature/hafs_moving_nest

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Add a new PBL diagnostic variable dkudiagnostic: atmosphere momentum diffusivity.
From @ChunxiZhang-NOAA and @AndrewHazelton

BinLiu-NOAA committed:
Update submodule ccpp/physics.

BinLiu-NOAA committed:
Merge branch 'support/HAFS' into feature/hafs_moving_nest
(Bill, Kyle, Xuejin, Bin)

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Add two ccpp suites: HAFS_v0_gfdlmp_tedmf_nonsst and HAFS_v0_gfdlmp_tedmf (Chunxi).

BinLiu-NOAA committed:
Update submodule ccpp/physics.

BinLiu-NOAA committed:
Update module_fcst_grid_comp.F90 from Rusty Benson to fix the warning message
of wrongly opened coupler.res files when having more than 3 nests.

BinLiu-NOAA committed:
Merge remote-tracking branch 'origin/feature/multi_nests' into feature/hafs_moving_nest

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/dkudiagnostic

BinLiu-NOAA committed:
Rename dkudiagnostic into dku and add dkt as a diagnostic variable.
Contributors: Andrew Hazelton, Chunxi Zhang, Bin Liu

BinLiu-NOAA committed:
Merge branch 'feature/dkudiagnostic' into support/HAFS

BinLiu-NOAA committed:
Update submodule ccpp/physics.

BinLiu-NOAA committed:
Merge branch 'feature/dkudiagnostic' into support/HAFS

BinLiu-NOAA committed:
Use unique standard_name for the new added dkt and dku diagnostic variables in
GFS_typedefs.meta.
Contributors: Chunxi Zhang and Bin Liu

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/dkudiagnostic' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafsv0.2_phase2' into support/HAFS

BinLiu-NOAA committed:
Point to support/HAFS branch of hafs-community fork of GFDL_atmos_cubed_sphere.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
*Unify the usage of Diag%dkt and Interstitial%dkt and change the standard_names
for dkt/dku back to atmosphere_heat/momentum_diffusivity.
*Update submodule ccpp/physics.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/dkudiagnostic' into support/HAFS

BinLiu-NOAA committed:
Change to allocate Diag%dkt/dku with Model%levs-1 vertical levels in GFS_typedefs.F90.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/dkudiagnostic' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Add a cplice namelist option to deal with the situation that HAFS FV3ATM-HYCOM
coupling without an active sea ice (e.g., CICE6) component. By default, the
cplice is true when cplflx is true (e.g., for the S2S application). However,
for the current HAFS FV3ATM-HYCOM coupling, cplice should set as false.  In the
future, it needs to be generalized/implemented to separate cplflx into cplocn
and cplice options.  (ChunxiZhang-NOAA and BinLiu-NOAA)

Commit:
fix for wave coupling with cmeps

Commit:
revert mods in cap and move it to ccpp level for now

Commit:
add white space again

BinLiu-NOAA committed:
Convert the parameter variables rlmx and elmx (maximum allowed mixing length
and dissipation mixing length in the GFS scale-aware TKE EDMF PBL scheme) into
namelist options. The default value for rlmx/elmx is 300.0 (unchanged). But
this allows developers to easily test rlmx/elmx with other values (e.g., using
100.0 in HAFS). These changes were made by ChunxiZhang-NOAA and BinLiu-NOAA
based on the rlmx/elmx value of 100.0 used by AndrewHazelton-NOAA.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_coupleww3

wramstrom committed:
Added call to perform moving nest at higher level, to allow access to physics variables.

BinLiu-NOAA committed:
*Update the interface and location to call the udpate_moving_nest.
*Update submodule atmos_cubed_sphere.
Note: Coding session among Bill, Xuejin, Kyle, Zhan, Chunxi, Weiguo, Bin

wramstrom committed:
Enable MOVING_NEST flag for this level of the build.  Correct module name for moving nest in atmos_model.F90.

wramstrom committed:
Added call for debugging output of moving nest variables to netCDF files.

BinLiu-NOAA committed:
*FV3 level changes needed for WeiguoWang-NOAA's modification of near-surface
mixing length in GFS sa-TKE EDMF PBL scheme. A new namelist option of sfc_rlm
was added to control this change, default value is 0 corresponding to no change.
The value of 1 will use zk for elm within the near-surface layer.
Contributors: WeiguoWang-NOAA, ChunxiZhang-NOAA, ZhanZhang-NOAA, BinLiu-NOAA

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_coupleww3

wramstrom committed:
Make GFS_init() public so that moving nest code can reset lat/lon values for physics arrays after nest move.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_sync_202107

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/hafs_sync_202107

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_coupleww3

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_sync_202107

BinLiu-NOAA committed:
Clean up and rename some HAFS related CCPP suites.

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/hafs_sync_202107

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_coupleww3

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_coupleww3

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_sync_202107

danrosen25 committed:
Update HAFS FV3 PR355 with change requests

* Remove merge_import and debug_merge from atmos_model_nml
* Add cpl_imp_mrg and cpl_imp_dbg to GFS_control (gfs_physics_nml)
* Remove diagnostic, replaced with existing cplprint_flag (DumpFields)
* Move write_grid_netcdf to wrt_fcst_grid in module_fcst_grid_comp.F90
* Update atmos_cubed_sphere submodule to match NOAA-GFDL/dev/emc

BinLiu-NOAA committed:
Update submodule GFDL_atmos_cubed_sphere.

BinLiu-NOAA committed:
Update submodule ccpp/physics.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Update .gitmodules and submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Update .gitmodules.

BinLiu-NOAA committed:
Add the cplocn2atm namelist option to turn on/off ocean feedback to FV3atm (#7)

Currently, when running coupled configurations/applications (e.g., HAFS or S2S). There is no available option to control one-way coupling to turn off the ocean model component feedback (e.g., SST) to FV3ATM.

Similar to the cplwav2atm namelist option, add a cplocn2atm namelist option to turn on/off the ocean model component feedback (e.g., SST) to the atmosphere model component. This is useful to control the one-way or two-way atmosphere-ocean coupling, as well as to configure different combinations for the atmosphere-ocean-wave coupling in HAFS and/or other applications.

BinLiu-NOAA committed:
*Add atmos_cubed_sphere/model/fv_tracker.F90 in CMakeLists.txt
*Update submodule atmos_cubed_sphere with FV3 internal tracker needed changes.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_moving_nest' into feature/hafs_vortex_tracking

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere and now feature/hafs_vortex_tracking can build successfully.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

DusanJovic-NOAA committed:
Update .gitmodules

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

DusanJovic-NOAA committed:
Create multiple fcstGrid objects, one for each fv3 grid

DusanJovic-NOAA committed:
Use pelist(1) as rootPet when broadcasting from top-level-domain

wramstrom committed:
Moved moving nest physics routines into separate file.  Combined moving nest logging and utils.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_nesting

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nesting
Working session among @wramstrom, @BinLiu-NOAA, @Xuejin-Zhang, @ChunxiZhang-NOAA, @A-Kyle.

BinLiu-NOAA committed:
Add module use line for mpp_set_current_pelist in module_fcst_grid_comp.F90.
Update submodule atmos_cubed_sphere.

wramstrom committed:
Enable 64 bit compiles

wramstrom committed:
Additional changes to permit 32bit/64bit compiles.

wramstrom committed:
Adapted moving nest code to handle new UFS weather model version of albedo and lake handling.

Commit:
Added t-dependent lon-lat (grid_mlont, grid_mlatt) to fv_diagnostics

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into support/HAFS

DusanJovic-NOAA committed:
Create multiple forecast grids

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into support/HAFS

wramstrom committed:
Added moving nest functionality for neststruct%i/j stagger.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/support/HAFS' into feature/hafs_nesting

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nesting

wramstrom committed:
Added preliminary land masked interpolation for moving nest.

Commit:
update submodule atmos_cubed_sphere.

wramstrom committed:
Regional Moving Nest Enabled. Fixed moving nest for land emissivity and land albedo. Static surface files renamed.

BinLiu-NOAA committed:
Merge commit '4545b4dfa0d20ce7106881b3d93d6e74b5e7b006' into feature/hafs_nesting

wramstrom committed:
Moving Nest updates for emis_lnd and emis_wat.

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into multiple_output_grids

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

DusanJovic-NOAA committed:
Update after merge

theurich committed:
Add missing error handling.

theurich committed:
No memory profiling by default (reduce log volume). Re-enable the GridCompDestroy() during Finalize().

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into multiple_output_grids

DusanJovic-NOAA committed:
Add support for multiple output grids

DusanJovic-NOAA committed:
Make app_domain an array (one element per output grid). Code cleanup.

DusanJovic-NOAA committed:
Remove some unused variables

DusanJovic-NOAA committed:
Remove 'app_domain' config parameter

BinLiu-NOAA committed:
Merge commit 'a9e79095120c90ea753d070d5d83dd01ac9464d4' into feature/hafs_nesting

wramstrom committed:
Added patch for nest motion of surface emissivity and albedo.

wramstrom committed:
Adding patch for surface emissivity for moving nest.
Merge branch 'feature/hafs_nesting' of https://github.com/hafs-community/fv3atm into feature/hafs_nesting

BinLiu-NOAA committed:
Merge branch 'feature/hafs_nesting' of https://github.com/hafs-community/fv3atm into feature/hafs_nesting

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

DusanJovic-NOAA committed:
Create global fcstGrid by assigning lat/lon coordinates instead of reading them from mosaic file

theurich committed:
Implement unified specification of output grids within model_configure. Fix a couple of comments.

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into multiple_output_grids

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/DJ/multiple_output_grids' into feature/hafs_nesting

DusanJovic-NOAA committed:
Change the name of the nest output files

BinLiu-NOAA committed:
Merge remote-tracking branch 'DJ/multiple_output_grids' into feature/hafs_nesting

DusanJovic-NOAA committed:
Use ESMF_GridCreateNoPeriDim for regional output grids

BinLiu-NOAA committed:
Merge remote-tracking branch 'DJ/multiple_output_grids' into feature/hafs_nest_sync

wramstrom committed:
Only enter moving nest code for moving nest configurations

wramstrom committed:
Only enter moving nest code for moving nest configurations.

Merge branch 'feature/hafs_nest_sync' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_sync

wramstrom committed:
Code cleanup of moving nest modules

DusanJovic-NOAA committed:
Use same global attributes for global_latlon and regional_latlon

danrosen25 committed:
Fix coupling pass-thru to fcstComp subdomain

* add input namelist option cpl_grid_id (1=parent)
* run state_diagnose for sparse fields (localDeCount.gt.0)
* move NUOPC_Advertise and NUOPC_Realize into fcstComp subdomain (a.k.a.
  'nests')
* move add grid coordinates and mask into SetServicesNest after grid
* create
* remove conditions for adding grid mask (previously no mask on nested
  configurations)

BinLiu-NOAA committed:
Merge remote-tracking branch 'DJ/multiple_output_grids' into feature/hafs_nest_sync

DusanJovic-NOAA committed:
Merge remote-tracking branch 'origin/develop' into multiple_output_grids

DusanJovic-NOAA committed:
Deallocate lon_start_wrtgrp/lon_end_wrtgrp arrays

DusanJovic-NOAA committed:
Remove ESMF_VMBarrier from ModelFinalize

wramstrom committed:
Added doxygen comments and regularized indentation.

wramstrom committed:
Added doxygen comments and regularized indentation.
Merge branch 'feature/hafs_nest_sync' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_sync

DusanJovic-NOAA committed:
Nullify local array (pelist) fcst_grid_comp before calling atmos_model_get_nth_domain_info()

DusanJovic-NOAA committed:
Update atmos_cubed_sphere

BinLiu-NOAA committed:
*Add the capability of downscaling the coupling variables (e.g., SST) from the
parent grid into the nested grid. This is needed since only the atmospheric
parent grid is coupled to other model components currently.
*Update the logic to determine the moving_nesting property.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/DJ/multiple_output_grids' into feature/hafs_nest_cpl

wramstrom committed:
Connected move_nsst to namelist option.

wramstrom committed:
Revised namelist options for move_nsst in moving nest.

wramstrom committed:
Moving nest for physics ifd only if using NSST.

BinLiu-NOAA committed:
Use moving_nest_parent and is_moving_nest to identify moving nest parent grid
and moving nest grid.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Minor correction, clean-up and simplification around domain specific advertise/realize. (#9)

Minor correction, clean-up and simplification around domain specific advertise/realize.
Co-authored-by: Gerhard Theurich <theurich@sourcespring.net>
Co-authored-by: Daniel Rosen <daniel.rosen@noaa.gov>

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

wramstrom committed:
Upgrade to read in monthly high-resolution albedo information for moving nest.

theurich committed:
Implement moving nest output. For moving domains, the Regrid() is
offloaded to the wrtComp. In that case use a simple Redist() to
move data, and coordinate updates from fcstComp to wrtComp. This
can be done with static RouteHandles, since the distribution does
not change. The wrtComp deals with the Regrid() problem, re-computing
new RouteHandles each output step for moving domains. The actual
output grid is static in this implementation.

wramstrom committed:
Handle moving nest interpolation of sfc roughness variables with land mask

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_cpl

wramstrom committed:
Reapply handling of NSST being turned off for coupled runs.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_nest_cpl' into feature/hafs_nest_cpl_moving

wramstrom committed:
Added Doxygen documentation and code cleanup.

Commit:
Add zsurf to diag fields for moving nest.

BinLiu-NOAA committed:
Add lon_cen and lat_cen (grid center longitude and latitude in radians) in
atmos_data_type, which will be updated when calling atmosphere_grid_ctr for the
moving nesting configuration. This is to support movable output grid for the
write grid component to ouput moving nest domain properly/efficiently.

wramstrom committed:
Updated formatting of header comments.

theurich committed:
Implement regridmethod specific tracing for moving nest regridding.

theurich committed:
Merge remote-tracking branch 'origin/feature/hafs_nest_cpl_movable_outgrid' into feature/hafs_nest_cpl_moving

DusanJovic-NOAA commented on atmos_model.F90:
```diff
@@ -101,6 +102,9 @@ module atmos_model_mod
 
+#ifdef MOVING_NEST
+use fv_moving_nest_main_mod, only: update_moving_nest, dump_moving_nest
+#endif MOVING_NEST
```
Please remove `MOVING_NEST` after `#endif`. GNU compiler complains about extra token at the end of #ifdef directive. Here and at two other places.

BinLiu-NOAA replied:
Thanks, @DusanJovic-NOAA! Done. Removed accordingly for both here and the other two locations.

BinLiu-NOAA committed:
Remove MOVING_NEST after #endif. GNU compiler complains about extra token at
the end of #ifdef directive. (Suggested by @DusanJovic-NOAA)

theurich committed:
Implement center coord determination for moving transferred grids.

DusanJovic-NOAA committed:
Add 'rotated_latlon_moving' output_grid (#10)

BinLiu-NOAA committed:
Revert "Add lon_cen and lat_cen (grid center longitude and latitude in radians) in atmos_data_type"

This reverts commit ead6f094d539a2d501ba519b20e52ab59cdcd903.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/origin/feature/hafs_nest_cpl' into feature/hafs_nest_cpl_movable_outgrid

BinLiu-NOAA committed:
Add 'regional_latlon_moving' output_grid type for moving nest.

BinLiu-NOAA committed:
Optimize the code changes related the regional/rotated_latlon_moving output
grid type in io/module_wrt_grid_comp.F90 (Based on suggestions from
@DusanJovic-NOAA).

BinLiu-NOAA committed:
Cosmetic indentation changes related to writ grid component support for moving nesting.

DusanJovic-NOAA committed:
Update grid specification for rotated_latlon_moving and regional_latlon_moving (#11)

* Update grid specification for rotated_latlon_moving and regional_latlon_moving

BinLiu-NOAA committed:
Fix the indentation of this lonPtr, latPtr line in io/module_wrt_grid_comp.F90.

danrosen25 commented on ccpp/data/GFS_typedefs.F90:
```diff
@@ -649,6 +649,7 @@ module GFS_typedefs
     logical              :: cplwav2atm      !< default no wav->atm coupling
     logical              :: cplchm          !< default no cplchm collection
     logical              :: use_cice_alb    !< default .false. - i.e. don't use albedo imported from the ice model
+    integer              :: cpl_grid_id     !< default grid 1 (parent)
```
I'd like to move this into the nems.configure file and attach it to the FV3 cap attributes instead of in the input.nml file.

BinLiu-NOAA replied:
@danrosen25, I think that's a good idea. Could you please help to make the needed code changes in a feature branch then merge it back into this feature/hafs_nest_cpl branch after testing? Thanks!

danrosen25 committed:
Move cpl_grid_id to FV3 model attributes
  * Set cpl_grid_id in nems.configure (default 1=parent)

wramstrom committed:
Update Coriolis terms after nest move.

wramstrom committed:
Moving nest coriolis terms.
Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Merge pull request #12 from hafs-community/feature/hafs_nest_cpl_fixcplgridid

Move cpl_grid_id to FV3 model attributes

wramstrom committed:
Simplified call to EOSHIFT for compiler portability.  Now omits BOUNDARY argument to use the default value of 0.0 (in single or double precision), instead of explicitly setting it to 0

wramstrom committed:
Merging EOSHIFT update for compiler portability.

Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Clean up trailing whitespace and update submodule atmos_cubed_sphere.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

wramstrom committed:
Modularization of moving nest and automated storm tracker.

wramstrom committed:
Modularization of moving nest and tracker.

Merge branch 'feature/hafs_nest_cpl' of https://github.com/hafs-community/fv3atm into feature/hafs_nest_cpl

wramstrom committed:
Separate moving_nest code into its own directory.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

wramstrom committed:
Made call to EOSHIFT portable to gfortran.

wramstrom committed:
Removed commented-out code from fv_control.F90.

wramstrom committed:
Revert logging output in fv_diagnostics.F90 to original format.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.

wramstrom committed:
Remove moving nest debugging statements from dycore core.

BinLiu-NOAA committed:
Revert back to the original forecast grid definition using the ESMF coord
dimensions so that the FV3-GOCART coupled regression tests can run through
successfuly. (From @junwang-noaa)

BinLiu-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Typo fix to correctly add @junwang-noaa's fix.
Amend the Revert back to the original forecast grid definition using the ESMF
coord dimensions so that the FV3-GOCART coupled regression tests can run
through successfuly. (From @junwang-noaa)

DusanJovic-NOAA commented on module_fcst_grid_comp.F90:
```diff
@@ -203,6 +224,94 @@ subroutine SetServicesNest(nest, rc)
       if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
     endif
 
+    ! - Create coordinate arrays around allocations held within Atmos data structure and set in Grid
+
+  ! call ESMF_GridGet(grid, staggerloc=ESMF_STAGGERLOC_CENTER, distgrid=distgrid, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! array = ESMF_ArrayCreate(distgrid, farray=Atmos%lon, indexflag=ESMF_INDEX_DELOCAL, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! call ESMF_GridSetCoord(grid, coordDim=1, staggerLoc=ESMF_STAGGERLOC_CENTER, array=array, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! array = ESMF_ArrayCreate(distgrid, farray=Atmos%lat, indexflag=ESMF_INDEX_DELOCAL, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! call ESMF_GridSetCoord(grid, coordDim=2, staggerLoc=ESMF_STAGGERLOC_CENTER, array=array, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! call ESMF_GridGet(grid, staggerloc=ESMF_STAGGERLOC_CORNER, distgrid=distgrid, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! array = ESMF_ArrayCreate(distgrid, farray=Atmos%lon_bnd, indexflag=ESMF_INDEX_DELOCAL, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! call ESMF_GridSetCoord(grid, coordDim=1, staggerLoc=ESMF_STAGGERLOC_CORNER, array=array, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! array = ESMF_ArrayCreate(distgrid, farray=Atmos%lat_bnd, indexflag=ESMF_INDEX_DELOCAL, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
+  ! call ESMF_GridSetCoord(grid, coordDim=2, staggerLoc=ESMF_STAGGERLOC_CORNER, array=array, rc=rc)
+  ! if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) return
+
```
Remove the commented-out code.

DusanJovic-NOAA commented on module_fcst_grid_comp.F90:
```diff
@@ -151,6 +161,11 @@ subroutine SetServicesNest(nest, rc)
     integer,dimension(2)  :: regdecomp                   !define delayout for the nest grid
     type(ESMF_Decomp_Flag):: decompflagPTile(2,6)
+    character(3)          :: myGridStr
+    type(ESMF_DistGrid)   :: distgrid
+    type(ESMF_Array)      :: array
```
Not needed anymore. Remove.

BinLiu-NOAA replied:
This comment is based on an outdated code. And with the latest code, these two lines are needed.

BinLiu-NOAA committed:
Update module_fcst_grid_comp.F90 to restore the support for moving-nesting output
grid. (@junwang-noaa and @BinLiu-NOAA)

wramstrom committed:
Moving nest bug fix for nest edge on prime meridian.

DusanJovic-NOAA committed:
Allocate lon_bnd, lat_bnd arrays to be the same size as coordinate distgrid

DusanJovic-NOAA committed:
Fix typos in comments

BinLiu-NOAA committed:
Ensure printing out the "PASS: fcstRUN phase ..." in the same line, so that one
can easily check/track the wallclock time for each time step.

BinLiu-NOAA committed:
Merge remote-tracking branch 'remotes/upstream/develop' into feature/hafs_nest_cpl

BinLiu-NOAA committed:
Indentation fix in fv3_cap.F90.

BinLiu-NOAA committed:
Ensure printing out the "PASS: fcstRUN phase ..." in the same line, so that one
can easily check/track the wallclock time for each time step. Make it work for
both gnu and intel compilers.

BinLiu-NOAA committed:
Update submodule atmos_cubed_sphere.
